Foundation/NSStringAPI.swift

```diff
@@ -759,6 +759,18 @@ extension StringProtocol where Index == String.Index {
     using encoding: String.Encoding,
     allowLossyConversion: Bool = false
   ) -> Data? {
+    let ephemeralString = _ephemeralString
+    if ephemeralString._core.isASCII {
+        switch encoding {
+                case .ascii: fallthrough
+                case .nonLossyASCII: fallthrough
+                case .utf8: return Data(
+                    bytes: UnsafeRawPointer(ephemeralString._core.startASCII), count: ephemeralString._core.count)
+                default:  return _ns.data(
+                    using: encoding.rawValue,
+                    allowLossyConversion: allowLossyConversion)
+        }
+    }
```

Comment:
This variant looks good to me; however we could do a touch better an get the unicode variation too.

```swift
public func data(using encoding: String.Encoding, allowLossyConversion: Bool = false) -> Data? {
        let ephemeralString = _ephemeralString
        if ephemeralString._core.isASCII {
            switch encoding {
            case .ascii: fallthrough
            case .nonLossyASCII: fallthrough
            case .utf8: return Data(
                bytes: UnsafeRawPointer(ephemeralString._core.startASCII), count: ephemeralString._core.count)
            default: break
            }
        } else {
            switch encoding {
            case .utf16: fallthrough
            case .unicode:
                return Data(bytes: UnsafeRawPointer(ephemeralString._core.startUTF16), count: ephemeralString._core.count * MemoryLayout<UInt16>.size)
            default: break
            }
        }
        
        return _ns.data(using: encoding.rawValue, allowLossyConversion: allowLossyConversion)
    }
```

Reply:
@phausler have tried implementing the code you suggested, but it doesn't seem to work. It looks like the conversion might be missing the byte order mark based on examining the data produced by the two code paths. Will have a look today into how the existing code does this to try and fix it.

Reply:
@phausler I've managed to get it to work using the following code, but it's not particularly neat! It basically creates the Byte Order Mark and prepends it to the front of the data.
```swift
public func data(
    using encoding: String.Encoding,
    allowLossyConversion: Bool = false
  ) -> Data? {
    let ephemeralString = _ephemeralString
    if ephemeralString._core.isASCII {
        switch encoding {
                case .ascii: fallthrough
                case .nonLossyASCII: fallthrough
                case .utf8: return Data(
                    bytes: UnsafeRawPointer(ephemeralString._core.startASCII), count: ephemeralString._core.count)
                default: break
        }
    }
    else {
        let str = ""
        var BOMData = str._ns.data(
           using: encoding.rawValue,
            allowLossyConversion: allowLossyConversion)
        
        switch encoding {
            case .utf16: fallthrough
            case .unicode:
                let bytes = UnsafeRawPointer(ephemeralString._core.startUTF16)
                let data = Data(
                    bytes: bytes, count: ephemeralString._core.count * MemoryLayout<UInt16>.size)
                BOMData!.append(data)
                let result = BOMData
                return result
            default: break
        }
    }
    return _ns.data(
      using: encoding.rawValue,
      allowLossyConversion: allowLossyConversion)
  }
```
Have pushed it to a [different branch](https://github.com/alexwishart/swift-corelibs-foundation/tree/unicode_fastpath) with the added tests and would appreciate some advice on how to potentially tidy it up. /cc @djones6 @ianpartridge